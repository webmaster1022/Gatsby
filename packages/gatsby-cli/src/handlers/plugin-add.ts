import reporter from "../reporter"
import { GatsbyPluginCreate, NPMPackageCreate } from "./plugin-add-utils"

const normalizePluginName = (plugin: string): string => {
  if (plugin.startsWith(`gatsby-`)) {
    return plugin
  }
  if (
    plugin.startsWith(`source-`) ||
    plugin.startsWith(`transformer-`) ||
    plugin.startsWith(`plugin-`)
  ) {
    return `gatsby-${plugin}`
  }
  return `gatsby-plugin-${plugin}`
}

async function installPluginPackage(
  plugin: string,
  root: string
): Promise<void> {
  const installTimer = reporter.activityTimer(`Installing ${plugin}`)

  installTimer.start()
  reporter.info(`Installing ${plugin}`)
  try {
    await NPMPackageCreate({ root, name: plugin })
    reporter.info(`Installed NPM package ${plugin}`)
  } catch (err) {
    reporter.error(JSON.parse(err)?.message)
    installTimer.setStatus(`FAILED`)
  }
  installTimer.end()
}

async function installPluginConfig(
  plugin: string,
  options: Record<string, unknown> | undefined,
  root: string
): Promise<void> {
  // Plugins can optionally include a key, to allow duplicates
  const [pluginName, pluginKey] = plugin.split(`:`)

  const installTimer = reporter.activityTimer(
    `Adding ${pluginName} ${pluginKey ? `(${pluginKey}) ` : ``}to gatsby-config`
  )

  installTimer.start()
  reporter.info(`Adding ${pluginName}`)
  try {
    await GatsbyPluginCreate({
      root,
      name: pluginName,
      options,
      key: pluginKey,
    })
    reporter.info(`Installed ${pluginName || pluginKey} in gatsby-config`)
  } catch (err) {
    reporter.error(JSON.parse(err)?.message)
    installTimer.setStatus(`FAILED`)
  }
  installTimer.end()
}

export async function addPlugins(
  plugins: Array<string>,
  pluginOptions: Record<string, Record<string, unknown>>,
  directory: string,
  packages: Array<string> = []
): Promise<void> {
  if (!plugins?.length) {
    reporter.error(`Please specify a plugin to install`)
    return
  }

  const pluginList = plugins.map(normalizePluginName)

  await Promise.all(
    packages.map(plugin => installPluginPackage(plugin, directory))
  )
  await Promise.all(
    pluginList.map(plugin =>
      installPluginConfig(plugin, pluginOptions[plugin], directory)
    )
  )
}
