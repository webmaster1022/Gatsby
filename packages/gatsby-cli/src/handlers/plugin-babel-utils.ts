import * as t from "@babel/types"
import generate from "@babel/generator"
import template from "@babel/template"
import { declare } from "@babel/helper-plugin-utils"
import type { ConfigAPI, PluginObj } from "@babel/core"

const getKeyNameFromAttribute = (node: any): any =>
  node.key.name || node.key.value

const unwrapTemplateLiteral = (str: string): string =>
  str.trim().replace(/^`/, ``).replace(/`$/, ``)

const isLiteral = (node: any): boolean =>
  t.isLiteral(node) || t.isStringLiteral(node) || t.isNumericLiteral(node)

const getObjectFromNode = (nodeValue: any): any => {
  if (!nodeValue || !nodeValue.properties) {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return getValueFromNode(nodeValue)
  }

  const props = nodeValue.properties.reduce((acc, curr) => {
    let value = null

    if (curr.value) {
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      value = getValueFromNode(curr.value)
    } else if (t.isObjectExpression(curr.value)) {
      value = curr.value.expression.properties.reduce((acc, curr) => {
        acc[getKeyNameFromAttribute(curr)] = getObjectFromNode(curr)
        return acc
      }, {})
    } else {
      throw new Error(`Did not recognize ${curr}`)
    }

    acc[getKeyNameFromAttribute(curr)] = value
    return acc
  }, {})

  return props
}

const getValueFromNode = (node: any): any => {
  if (t.isTemplateLiteral(node)) {
    // @ts-ignore - fix me
    delete node.leadingComments
    // @ts-ignore - fix me
    delete node.trailingComments
    // @ts-ignore - fix me
    const literalContents = generate(node).code
    return unwrapTemplateLiteral(literalContents)
  }

  if (isLiteral(node)) {
    return node.value
  }

  if (node.type === `ArrayExpression`) {
    return node.elements.map(getObjectFromNode)
  }

  if (node.type === `ObjectExpression`) {
    return getObjectFromNode(node)
  }

  return null
}

function isDefaultExport(node): boolean {
  if (!node || !t.isMemberExpression(node)) {
    return false
  }

  const { object, property } = node

  if (!t.isIdentifier(object) || object.name !== `module`) return false
  if (!t.isIdentifier(property) || property.name !== `exports`) return false

  return true
}

const getOptionsForPlugin = (node: any): any => {
  if (!t.isObjectExpression(node) && !t.isLogicalExpression(node)) {
    return undefined
  }

  let options

  // When a plugin is added conditionally with && {}
  if (t.isLogicalExpression(node)) {
    // @ts-ignore - fix me
    options = node.right.properties.find(
      property => property.key.name === `options`
    )
  } else {
    // @ts-ignore - fix me
    options = node.properties.find(property => property.key.name === `options`)
  }

  if (options) {
    return getObjectFromNode(options.value)
  }

  return undefined
}

const getKeyForPlugin = (node: any): any => {
  if (t.isObjectExpression(node)) {
    // @ts-ignore - fix me
    const key = node.properties.find(p => p.key.name === `__key`)

    // @ts-ignore - fix me
    return key ? getValueFromNode(key.value) : null
  }

  // When a plugin is added conditionally with && {}
  if (t.isLogicalExpression(node)) {
    // @ts-ignore - fix me
    const key = node.right.properties.find(p => p.key.name === `__key`)

    return key ? getValueFromNode(key.value) : null
  }

  return null
}

const getNameForPlugin = (node: any): any => {
  if (t.isStringLiteral(node) || t.isTemplateLiteral(node)) {
    return getValueFromNode(node)
  }

  if (t.isObjectExpression(node)) {
    // @ts-ignore - fix me
    const resolve = node.properties.find(p => p.key.name === `resolve`)

    // @ts-ignore - fix me
    return resolve ? getValueFromNode(resolve.value) : null
  }

  // When a plugin is added conditionally with && {}
  if (t.isLogicalExpression(node)) {
    // @ts-ignore - fix me
    const resolve = node.right.properties.find(p => p.key.name === `resolve`)

    return resolve ? getValueFromNode(resolve.value) : null
  }

  return null
}

const getPlugin = (node: any): any => {
  const plugin = {
    name: getNameForPlugin(node),
    options: getOptionsForPlugin(node),
  }

  const key = getKeyForPlugin(node)

  if (key) {
    return { ...plugin, key }
  }

  return plugin
}

function buildPluginNode({ name, options, key }): any {
  if (!options && !key) {
    return t.stringLiteral(name)
  }

  const pluginWithOptions = template(
    `
    const foo = {
      resolve: '${name}',
      options: ${JSON.stringify(options, null, 2)},
      ${key ? `__key: "` + key + `"` : ``}
    }
  `,
    { placeholderPattern: false }
  )()

  // @ts-ignore - fix me
  return pluginWithOptions.declarations[0].init
}

function addPluginsToConfig({
  pluginNodes,
  pluginOrThemeName,
  options,
  key,
}): void {
  if (t.isCallExpression(pluginNodes.value)) {
    const plugins = pluginNodes.value.callee.object.elements.map(getPlugin)
    const matches = plugins.filter(plugin => {
      if (!key) {
        return plugin.name === pluginOrThemeName
      }

      return plugin.key === key
    })

    if (!matches.length) {
      const pluginNode = buildPluginNode({
        name: pluginOrThemeName,
        options,
        key,
      })

      pluginNodes.value.callee.object.elements.push(pluginNode)
    } else {
      pluginNodes.value.callee.object.elements =
        pluginNodes.value.callee.object.elements.map(node => {
          const plugin = getPlugin(node)

          if (plugin.key !== key) {
            return node
          }

          if (!plugin.key && plugin.name !== pluginOrThemeName) {
            return node
          }

          return buildPluginNode({
            name: pluginOrThemeName,
            options,
            key,
          })
        })
    }
  } else {
    const plugins = pluginNodes.value.elements.map(getPlugin)
    const matches = plugins.filter(plugin => {
      if (!key) {
        return plugin.name === pluginOrThemeName
      }

      return plugin.key === key
    })

    if (!matches.length) {
      const pluginNode = buildPluginNode({
        name: pluginOrThemeName,
        options,
        key,
      })

      pluginNodes.value.elements.push(pluginNode)
    } else {
      pluginNodes.value.elements = pluginNodes.value.elements.map(node => {
        const plugin = getPlugin(node)

        if (plugin.key !== key) {
          return node
        }

        if (!plugin.key && plugin.name !== pluginOrThemeName) {
          return node
        }

        return buildPluginNode({
          name: pluginOrThemeName,
          options,
          key,
        })
      })
    }
  }
}

function getPluginNodes(
  properties: Array<t.ObjectProperty | t.ObjectMethod | t.SpreadElement>
): t.ObjectProperty | undefined {
  return properties.find(
    prop =>
      t.isObjectProperty(prop) &&
      t.isIdentifier(prop.key) &&
      prop.key.name === `plugins`
  ) as t.ObjectProperty
}

/**
 * Insert plugins selected in create-gatsby questionnaire into `gatsby-config` files.
 *
 * Scope is limited to the `gatsby-config` files in `gatsby-starter-minimal` and
 * `gatsby-starter-minimal-ts`. Does not support general usage with other `gatsby-config` files.
 * Changes to the config object in those files may require a change to this transformer.
 *
 * @see {@link https://github.com/gatsbyjs/gatsby/blob/master/starters/gatsby-starter-minimal/gatsby-config.js}
 * @see {@link https://github.com/gatsbyjs/gatsby/blob/master/starters/gatsby-starter-minimal-ts/gatsby-config.ts}
 */
export default declare(
  (
    api: ConfigAPI,
    args: {
      pluginOrThemeName: string
      options: unknown
      key: string
    }
  ): PluginObj => {
    api.assertVersion(7)

    return {
      visitor: {
        /**
         * Handle `module.exports = { ..., plugins: [] }` from `gatsby-config.js` in `gatsby-starter-minimal`.
         * @see {@link https://github.com/gatsbyjs/gatsby/blob/master/starters/gatsby-starter-minimal/gatsby-config.js}
         */
        ExpressionStatement(path): void {
          const { node } = path
          if (!t.isAssignmentExpression(node.expression)) {
            return
          }

          const { left, right } = node.expression
          if (!isDefaultExport(left) || !t.isObjectExpression(right)) {
            return
          }

          const pluginNodes = getPluginNodes(right.properties)

          if (
            !t.isObjectProperty(pluginNodes) ||
            !t.isArrayExpression(pluginNodes.value)
          ) {
            return
          }

          addPluginsToConfig({ pluginNodes, ...args })

          path.stop()
        },

        /**
         * Handle `const config = { ..., plugins: [] }; export default config` in `gatsby-config.ts` in `gatsby-starter-minimal-ts`.
         * @see {@link https://github.com/gatsbyjs/gatsby/blob/master/starters/gatsby-starter-minimal-ts/gatsby-config.ts}
         */
        VariableDeclaration(path): void {
          const { node } = path
          const configDeclaration = node.declarations.find(
            dec =>
              t.isIdentifier(dec.id) && dec.id.name === `config` && dec.init
          )
          const config = configDeclaration?.init
          if (!t.isObjectExpression(config)) {
            return
          }

          const pluginNodes = getPluginNodes(config.properties)

          if (
            !t.isObjectProperty(pluginNodes) ||
            !t.isArrayExpression(pluginNodes.value)
          ) {
            return
          }

          addPluginsToConfig({ pluginNodes, ...args })

          path.stop()
        },
      },
    }
  }
)
