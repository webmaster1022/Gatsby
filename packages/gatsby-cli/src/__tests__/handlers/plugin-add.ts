import { ensureDir, copyFile, readFile, rm } from "fs-extra"
import { join } from "path"
import { addPlugins } from "../../handlers/plugin-add"

/**
 * Copy files from minimal starters instead of testing against static gatsby-configs
 * in fixtues so that these break if we change the starter configs in a breaking way.
 *
 * Not using `jest.each` since I find that much harder to read and debug.
 * @see {@link https://jestjs.io/docs/api#testeachtablename-fn-timeout}
 */

const root = join(__dirname, `../../../../..`)
const fixtures = join(__dirname, `../fixtures`)
const config = {
  js: {
    starter: `${root}/starters/gatsby-starter-minimal/gatsby-config.js`,
    fixture: `${fixtures}/gatsby-config.js`,
  },
  ts: {
    starter: `${root}/starters/gatsby-starter-minimal-ts/gatsby-config.ts`,
    fixture: `${fixtures}/gatsby-config.ts`,
  },
}
const plugin = {
  hello: `gatsby-plugin-hello`,
  world: `gatsby-plugin-world`,
}

describe(`addPlugins`, () => {
  beforeAll(async () => {
    await ensureDir(fixtures)
  })

  describe(`gatsby-config.js`, () => {
    beforeEach(async () => {
      await copyFile(config.js.starter, config.js.fixture)
    })

    afterEach(async () => {
      await rm(config.js.fixture)
    })

    it(`should not write with no plugins`, async () => {
      await addPlugins([], {}, fixtures, [])
      const gatsbyConfig = (await readFile(config.js.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })

    it(`should write a single plugin`, async () => {
      await addPlugins([plugin.hello], {}, fixtures, [])
      const gatsbyConfig = (await readFile(config.js.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })

    it(`should write multiple plugins`, async () => {
      await addPlugins([plugin.hello, plugin.world], {}, fixtures, [])
      const gatsbyConfig = (await readFile(config.js.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })

    it(`should write a plugin with options`, async () => {
      await addPlugins(
        [plugin.hello],
        {
          [plugin.hello]: {
            greet: true,
          },
        },
        fixtures,
        []
      )
      const gatsbyConfig = (await readFile(config.js.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })
  })

  describe(`gatsby-config.ts`, () => {
    beforeEach(async () => {
      await copyFile(config.ts.starter, config.ts.fixture)
    })

    afterEach(async () => {
      await rm(config.ts.fixture)
    })

    it(`should not write with no plugins`, async () => {
      await addPlugins([], {}, fixtures, [])
      const gatsbyConfig = (await readFile(config.ts.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })

    it(`should write a single plugin`, async () => {
      await addPlugins([plugin.hello], {}, fixtures, [])
      const gatsbyConfig = (await readFile(config.ts.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })

    it(`should write multiple plugins`, async () => {
      await addPlugins([plugin.hello, plugin.world], {}, fixtures, [])
      const gatsbyConfig = (await readFile(config.ts.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })

    it(`should write a plugin with options`, async () => {
      await addPlugins(
        [plugin.hello],
        {
          [plugin.hello]: {
            greet: true,
          },
        },
        fixtures,
        []
      )
      const gatsbyConfig = (await readFile(config.ts.fixture)).toString()
      expect(gatsbyConfig).toMatchSnapshot()
    })
  })
})
