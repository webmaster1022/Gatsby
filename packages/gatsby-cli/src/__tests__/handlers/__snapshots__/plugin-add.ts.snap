// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPlugins gatsby-config.js should not write with no plugins 1`] = `
"module.exports = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`,
  },
  plugins: [],
}
"
`;

exports[`addPlugins gatsby-config.js should write a plugin with options 1`] = `
"module.exports = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`
  },
  plugins: [{
    resolve: 'gatsby-plugin-hello',
    options: {
      \\"greet\\": true
    }
  }]
};"
`;

exports[`addPlugins gatsby-config.js should write a single plugin 1`] = `
"module.exports = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`
  },
  plugins: [\\"gatsby-plugin-hello\\"]
};"
`;

exports[`addPlugins gatsby-config.js should write multiple plugins 1`] = `
"module.exports = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`
  },
  plugins: [\\"gatsby-plugin-hello\\", \\"gatsby-plugin-world\\"]
};"
`;

exports[`addPlugins gatsby-config.ts should not write with no plugins 1`] = `
"import type { GatsbyConfig } from \\"gatsby\\"

const config: GatsbyConfig = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`,
  },
  // More easily incorporate content into your pages through automatic TypeScript type generation and better GraphQL IntelliSense.
  // If you use VSCode you can also use the GraphQL plugin
  // Learn more at: https://gatsby.dev/graphql-typegen
  graphqlTypegen: true,
  plugins: [],
}

export default config
"
`;

exports[`addPlugins gatsby-config.ts should write a plugin with options 1`] = `
"import type { GatsbyConfig } from \\"gatsby\\";

const config: GatsbyConfig = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`
  },
  // More easily incorporate content into your pages through automatic TypeScript type generation and better GraphQL IntelliSense.
  // If you use VSCode you can also use the GraphQL plugin
  // Learn more at: https://gatsby.dev/graphql-typegen
  graphqlTypegen: true,
  plugins: [{
    resolve: 'gatsby-plugin-hello',
    options: {
      \\"greet\\": true
    }
  }]
};

export default config;
"
`;

exports[`addPlugins gatsby-config.ts should write a single plugin 1`] = `
"import type { GatsbyConfig } from \\"gatsby\\";

const config: GatsbyConfig = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`
  },
  // More easily incorporate content into your pages through automatic TypeScript type generation and better GraphQL IntelliSense.
  // If you use VSCode you can also use the GraphQL plugin
  // Learn more at: https://gatsby.dev/graphql-typegen
  graphqlTypegen: true,
  plugins: [\\"gatsby-plugin-hello\\"]
};

export default config;
"
`;

exports[`addPlugins gatsby-config.ts should write multiple plugins 1`] = `
"import type { GatsbyConfig } from \\"gatsby\\";

const config: GatsbyConfig = {
  siteMetadata: {
    siteUrl: \`https://www.yourdomain.tld\`
  },
  // More easily incorporate content into your pages through automatic TypeScript type generation and better GraphQL IntelliSense.
  // If you use VSCode you can also use the GraphQL plugin
  // Learn more at: https://gatsby.dev/graphql-typegen
  graphqlTypegen: true,
  plugins: [\\"gatsby-plugin-hello\\", \\"gatsby-plugin-world\\"]
};

export default config;
"
`;
